src/cpr_data_access/parser_models.py

```diff
@@ -180,21 +194,82 @@ def check_html_pdf_metadata(cls, values):
         return values
 
     @property
-    def text_blocks(self) -> Sequence[TextBlock]:  # type: ignore
+    def text_blocks(self) -> Sequence[TextBlock]:
         """
-        Return the text blocks in the document. These could differ in format depending on the content type.
+        Return the text blocks in the document.
+
+        These could differ in format depending on the content type.
 
         :return: Sequence[TextBlock]
         """
-
         if self.document_content_type == CONTENT_TYPE_HTML:
-            return self.html_data.text_blocks  # type: ignore
+            html_data: HTMLData = self.html_data
+            return html_data.text_blocks
         elif self.document_content_type == CONTENT_TYPE_PDF:
-            return self.pdf_data.text_blocks  # type: ignore
+            pdf_data: PDFData = self.pdf_data
+            return pdf_data.text_blocks
+        return []
 
     def to_string(self) -> str:  # type: ignore
         """Return the text blocks in the parser output as a string"""
 
         return " ".join(
             [text_block.to_string().strip() for text_block in self.text_blocks]
         )
+
+    def detect_and_set_languages(self) -> "ParserOutput":
+        """
+        Detect language of the text and set the language attribute.
+
+        Return an instance of ParserOutput with the language attribute set. Assumes
+        that a document only has one language.
+        """
+
+        if self.document_content_type != CONTENT_TYPE_HTML:
+            logger.warning(
+                "Language detection should not be required for non-HTML documents, "
+                "but it has been run on one. This will overwrite any document "
+                "languages detected via other means, e.g. OCR. "
+            )
+
+        # language detection is not deterministic, so we need to set a seed
+        DetectorFactory.seed = 0
+
+        if len(self.text_blocks) > 0:
+            detected_language = detect(self.to_string())
```

Comment:
Do we need handle any errors? Just been looking here: https://github.com/shuyo/language-detection/blob/master/src/com/cybozu/labs/langdetect/LangDetectException.java

Reply:
Ah yeah - see here: https://github.com/Mimino666/langdetect/blob/a1598f1afcbfe9a758cfd06bd688fbc5780177b2/langdetect/detector.py#L150

Reply:
So you're saying we need to catch the following exception: 

`LangDetectException`

Reply:
Yeah - probably best to handle it here so we can guarantee the behaviour - unless you are ok with the exception bubbling up?

Reply:
No I'll add that in, this method is the one from the main parser so lucky if we haven't been caught by it thus far!

Reply:
Co-pilot scares me sometimes, how does it know! 

![image](https://github.com/climatepolicyradar/data-access/assets/58440325/4fd7ba84-7d4d-468a-9983-f94a6156ed1e)